performance.md

```diff
@@ -0,0 +1,341 @@
+; Size: 142 bytes. Origin: #x52D4815D
+~~~
+
+You may specify different safety and speed level and perform optimization
+strategy globally.
+
+### Code Inline
+
+The declaration [`inline`][inline] replaces function calls with function
+body. It will save the cost of function calls but will potentially increase
+the code size. The best situation to use `inline` might be those small but
+frequently used functions. The following snippet shows how to encourage and
+prohibit code inline.
+
+~~~lisp
+;; The globally defined function DISPATCH should be open-coded,
+;; if the implementation supports inlining, unless a NOTINLINE 
+;; declaration overrides this effect.
+(declaim (inline dispatch))
+(defun dispatch (x) (funcall (get (car x) 'dispatch) x))
+
+;; Here is an example where inlining would be encouraged.
+(defun top-level-1 () (dispatch (read-command)))
+ 
+;; Here is an example where inlining would be prohibited.
+(defun top-level-2 ()
+  (declare (notinline dispatch))
+  (dispatch (read-command)))
+
+;; Here is an example where inlining would be prohibited.
+(declaim (notinline dispatch))
+(defun top-level-3 () (dispatch (read-command)))
+
+;; Here is an example where inlining would be encouraged.
+(defun top-level-4 () 
+  (declare (inline dispatch))
+  (dispatch (read-command)))
+~~~
+
+## Optimizing Generic Functions
+
+### Using Static Dispatch
+
+Generic functions provide much convenience and flexibility during
+development. However, the flexibility comes with cost: generic methods are
+much slower than trivial functions. The performance cost becomes a burden
+especially when the flexibility is not needed.
+
+Package [`inlined-generic-function`][inlined-generic-function] provides
+functions to convert generic functions to static dispatch, moving dispatch
+cost to compile-time. You just need to define generic function as a
+`inlined-generic-function`.
+
+~~~lisp
+* (defgeneric plus (a b)
+    (:generic-function-class inlined-generic-function))
+#<INLINED-GENERIC-FUNCTION HELLO::PLUS (2)>
+
+* (defmethod plus ((a fixnum) (b fixnum))
+    (+ a b))
+#<INLINED-METHOD HELLO::PLUS (FIXNUM FIXNUM) {10056D7513}>
+
+* (defun func-using-plus (a b)
+    (plus a b))
+FUNC-USING-PLUS
+
+* (defun func-using-plus-inline (a b)
+    (declare (inline plus))
+    (plus a b))
+FUNC-USING-PLUS-INLINE
+
+* (time
+   (dotimes (i 100000)
+     (func-using-plus 100 200)))
+Evaluation took:
+  0.018 seconds of real time
+  0.017819 seconds of total run time (0.017800 user, 0.000019 system)
+  100.00% CPU
+  3 lambdas converted
+  71,132,440 processor cycles
+  6,586,240 bytes consed
+
+* (time
+   (dotimes (i 100000)
+     (func-using-plus-inline 100 200)))
+Evaluation took:
+  0.001 seconds of real time
+  0.000326 seconds of total run time (0.000326 user, 0.000000 system)
+  0.00% CPU
+  1,301,040 processor cycles
+  0 bytes consed
+~~~
+
+The inlining is not enabled by default because once inlined, changes made to
+methods will not be reflected.
```

Comment:
That is annoying. What can we do then ? (I suppose re-compiling the function containing the inlined method, as with macros ?)

Reply:
I do not think that there is anything we can do now.

During development of course you can re-compile the code, but IMHO the library is designed for deployment, where it requires more advanced techniques to update code changes.